### ISSUE TITLE
Add `dim` to `pytorch_lightning.metrics.PSNR`

### ISSUE THREAD

**Issue by manipopopo** [PyTorchLightning/pytorch-lightning#5933]
## 🚀 Feature

Add `dim` to `pytorch_lightning.metrics.PSNR` so that users can specify the dimension (or dimensions) for the mean squared error reduction if _PSNR_ computation.

### Motivation

Suppose we have two image pairs

```python
(pred_image_1, target_image_1)
(pred_image_2, target_image_2)
```

`pytorch_lightning.metrics.PSNR` computes the _PSNR_ as

```python
squared_error_1 = (pred_image_1 - target_image_1) ** 2
squared_error_2 = (pred_image_2 - target_image_2) ** 2
mean_squared_error = (squared_error_1 + squared_error_2) / (squared_error_1.numel() + squared_error_2.numel())
psnr = -10.0 * log(mean_squared_error)
```

If `pred_image_1.numel()` >> `pred_image_2.numel()`, the quality of `pred_image_1` may bias the output `pnsr`.
It will be helpful if we can compute the `psnr`s of each image pair separately and average `psnr`s:

```python
squared_error_1 = (pred_image_1 - target_image_1) ** 2
psnr_1 = -10.0 * log(squared_error_1.mean())
squared_error_2 = (pred_image_2 - target_image_2) ** 2
psnr_2 = -10.0 * log(squared_error_2.mean())
psnr = (psnr_1 + psnr_2) / 2
```

### Pitch

Add `dim` to `pytorch_lightning.metrics.PSNR`:

```python
class PSNR(Metric):
    def __init__(
        self,
        data_range: Optional[float] = None,
        base: float = 10.0,
        dim: Union[int, Sequence[int]] = (),  # New.
        reduction: str = 'elementwise_mean',  # Unused?
        compute_on_step: bool = True,
        dist_sync_on_step: bool = False,
        process_group: Optional[Any] = None,
    ):
        # ...

        self._dim = tuple(dim) if isinstance(dim, Sequence) else dim
        self.add_state("sum_psnr", default=torch.tensor(0.0), dist_reduce_fx="sum")
        self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
        # ...

    def update(self, preds: torch.Tensor, target: torch.Tensor):
        psnrs = compute_psnr(preds, target, dim=self._dim) # New.
        self.psnrs += psnrs.sum()
        self.total += psnrs.numel()

    def compute(self):
        return self.psnrs / self.total


preds_nchw = torch.rand([32, 3, 224, 224]
targets_nchw = torch.rand([32, 3, 224, 224]
metric = PSNR(dim=(1, 2, 3))
average_psnr_of_n_images = metric(preds_nchw, targets_nchw)
```

### Alternatives

If the original behavior should be kept, maybe we can do

```python
class PSNR(Metric):
    def __init__(
        self,
        data_range: Optional[float] = None,
        base: float = 10.0,
        dim: Optional[Union[int, Sequence[int]]] = None,  # `None` for original behavior
        reduction: str = 'elementwise_mean',  # Unused?
        compute_on_step: bool = True,
        dist_sync_on_step: bool = False,
        process_group: Optional[Any] = None,
    ):
        # ...

        self._dim = tuple(dim) if isinstance(dim, Sequence) else dim
        if self._dim is None:
            # Original behavior.
            self.add_state("sum_squared_error", default=torch.tensor(0.0), dist_reduce_fx="sum")
        else:
            self.add_state("sum_psnr", default=torch.tensor(0.0), dist_reduce_fx="sum")
        self.add_state("total", default=torch.tensor(0), dist_reduce_fx="sum")
        # ...
```

**Comment by SkafteNicki** [comment]
Hi @manipopopo,
Thanks for your suggestion. I think it is a good idea to add a `dim` argument to PSNR. Would you be up for sending a PR with the enhancement?

**Comment by manipopopo** [comment]
Hi @SkafteNicki , I’d be happy to. Maybe I can send a PR tomorrow.

**Comment by SkafteNicki** [comment]
@manipopopo sound good to me :]
Please ping me in your PR.
